snuba/cli/__init__.py

```diff
@@ -1,51 +1,68 @@
-logger = logging.getLogger("snuba_init")
 
-start = time.perf_counter()
-logger.info("Initializing Snuba...")
-metrics = MetricsWrapper(environment_metrics, "cli")
+with sentry_sdk.start_transaction(
+    op="snuba_init", name="Snuba CLI Initialization", sampled=True
+):
 
+    setup_logging("DEBUG")
+    logger = logging.getLogger("snuba_init")
 
-@click.group()
-@click.version_option()
-def main() -> None:
-    """\b
-                         o
-                        O
-                        O
-                        o
-    .oOo  'OoOo. O   o  OoOo. .oOoO'
-    `Ooo.  o   O o   O  O   o O   o
-        O  O   o O   o  o   O o   O
-    `OoO'  o   O `OoO'o `OoO' `OoO'o"""
+    start = time.perf_counter()
+    logger.info("Initializing Snuba CLI...")
+    metrics = MetricsWrapper(environment_metrics, "cli")
 
+    plugin_folder = os.path.dirname(__file__)
 
-with sentry_sdk.start_transaction(
-    op="snuba_init", name="Snuba CLI Initialization", sampled=True
-):
-    for loader, module_name, is_pkg in walk_packages(__path__, __name__ + "."):
-        logger.info(f"Loading module {module_name}")
-        with sentry_sdk.start_span(op="import", description=module_name):
-            module = __import__(module_name, globals(), locals(), ["__name__"])
-            cmd = getattr(module, module_name.rsplit(".", 1)[-1])
-            if isinstance(cmd, click.Command):
-                main.add_command(cmd)
-
-init_time = time.perf_counter() - start
-metrics.gauge("snuba_init", init_time)
-logger.info(f"Snuba initialization took {init_time}s")
-
-structlog.reset_defaults()
+    class SnubaCLI(click.MultiCommand):
+        def list_commands(self, ctx: Any) -> list[str]:
+            rv = []
+            for filename in os.listdir(plugin_folder):
+                if filename.endswith(".py") and filename != "__init__.py":
+                    rv.append(filename[:-3])
+            rv.sort()
+            return rv
+
+        def get_command(self, ctx: Any, name: str) -> click.Command:
+            logger.info(f"Loading command {name}")
+            with sentry_sdk.start_span(op="import", description=name):
+                ns: dict[str, click.Command] = {}
+                fn = os.path.join(plugin_folder, name + ".py")
+                with open(fn) as f:
+                    code = compile(f.read(), fn, "exec")
+                    eval(code, ns, ns)
+                initialize.initialize()
+                return ns[name]
+
+    @click.command(cls=SnubaCLI)
+    def main() -> None:
+        """\b
+                             o
+                            O
+                            O
+                            o
+        .oOo  'OoOo. O   o  OoOo. .oOoO'
+        `Ooo.  o   O o   O  O   o O   o
+            O  O   o O   o  o   O o   O
+        `OoO'  o   O `OoO'o `OoO' `OoO'o"""
+
+    init_time = time.perf_counter() - start
+    metrics.gauge("snuba_init", init_time)
+    logger.info(f"Snuba initialization took {init_time}s")
```

Comment:
Is this still true? I don't think  `init_time` is accurate anymore.

Reply:
fixed